### PR TITLE
fix(cl): per-segment Uniswap v3 attribution for staked reserve deltas (#666)

### LINKED ISSUE #666: bug(cl-staked): negative stakedReserve0/1 running counters on 166 CL pools across all chains
## Summary

166 CL pools across every Velodrome chain have `stakedReserve0` or `stakedReserve1` running counters that have drifted **negative** over their lifetime. The fields are typed `BigInt @config(precision: 76)` and represent staked reserves in raw token units — they should never go below zero.

The defensive clamp in `src/Aggregators/LiquidityPoolAggregator.ts:444–457` (`stakedReserve0 > 0n ? stakedReserve0 : 0n`) masks this at the USD layer, so `currentLiquidityStakedUSD` reads as $0 (not negative), but the raw `stakedReserve0/1` fields persist as negative and are queryable.

## Distribution

Counts of pools where `stakedReserve0 < 0` OR `stakedReserve1 < 0` (deployed indexer at `https://indexer.us.hyperindex.xyz/3d71402/v1/graphql`):

| chainId | chain      | pools |
|---------|------------|-------|
| 10      | Optimism   | 69    |
| 1868    | Soneium    | 18    |
| 57073   | Ink        | 15    |
| 1135    | Lisk       | 13    |
| 1923    | Swell      | 11    |
| 130     | Unichain   | 8     |
| 5330    | Superseed  | 8     |
| 252     | Fraxtal    | 6     |
| 1750    | Metal      | 6     |
| 34443   | Mode       | 6     |
| 42220   | Celo       | 6     |
| **total** | | **166** |

All affected pools are CL (`isCL: true`).

## Example pools

```
chain=10 pool=10-0xCAfFfE7… cl=true sR0=-252490000 sR1=2727052158459901178 swaps=1758
chain=10 pool=10-0xA61FBA4… cl=true sR0=127708019961736589974376 sR1=-135152857691846445422805 swaps=2597
chain=10 pool=10-0x8E55FF6… cl=true sR0=69892965020 sR1=-372720456187214354447837 swaps=6715
chain=10 pool=10-0x20efb6B… cl=true sR0=-205923431041 sR1=624408760215215304960574 swaps=4825
chain=10 pool=10-0xE96CEd7… cl=true sR0=-1132972076516482452154 sR1=1038943274066645804743 swaps=301
chain=10 pool=10-0x8Fc5014… cl=true sR0=-4350338725843752391483 sR1=4343863089208118709421 swaps=2099
chain=130 pool=130-0x9D9eE4… cl=true sR0=1288655334 sR1=-1200618 swaps=34409
chain=10 pool=10-0xa7F59fB… cl=true sR0=9387630595 sR1=-205570844507553670106 swaps=2523
chain=10 pool=10-0xed6b1E4… cl=true sR0=-6329965150454070643652 sR1=16990213525831706438126 swaps=848
chain=10 pool=10-0xB7Dc583… cl=true sR0=754439645137341967972400 sR1=-2204887454932230742036463468 swaps=54830
```

## Likely root cause

The reserve accumulation path is in `src/EventHandlers/Gauges/GaugeSharedLogic.ts:130-133` (running counter) and `src/Aggregators/LiquidityPoolAggregator.ts:385-387` (incremental diff). The incremental diff source for staked reserves is `processTickCrossingsForStaked` plus gauge deposit/withdraw flows.

Hypotheses to investigate (in rough likelihood order):

1. **Tick-crossing direction sign error** — when a swap crosses a staked tick, the contribution direction (add vs. subtract liquidity) may be flipped in some swap-direction × tick-position combinations.
2. **Stake/unstake ordering with tick crossings** — if a `Deposit` or `Withdraw` event is processed in the wrong order relative to swap tick crossings within the same block, the running counter loses an event boundary.
3. **`stakedTickEdges` sparse-list invariant mismatch** — the recent refactor (`#653` / `7ad861e`) introduced "log+drop invariant mismatch instead of throwing" for stakedTickEdges. Drops may be silently corrupting the running accumulator.
4. **Initial-state computation off-by-one** — when `hasStakes` first latches true, the initial L→reserve conversion may not use the sqrtPrice / tick that matches the next swap.

The defensive clamp at `LiquidityPoolAggregator.ts:444–457` was likely added to mask symptom; root cause was never tracked.

## Reproduction (read-only)

```graphql
query {
  LiquidityPoolAggregator(
    where: {
      _or: [
        { stakedReserve0: { _lt: "0" } }
        { stakedReserve1: { _lt: "0" } }
      ]
    }
    limit: 10
  ) {
    id
    chainId
    isCL
    stakedReserve0
    stakedReserve1
    currentLiquidityStaked
    currentLiquidityStakedUSD
    numberOfSwaps
  }
}
```

## Acceptance

- [ ] Root-cause the drift source (tick crossings vs gauge events vs sparse-list invariant)
- [ ] Add a regression test that exercises the failing code path on the smallest reproducer pool from the list above (e.g. `10-0xE96CEd7…` with 301 swaps is a candidate)
- [ ] Remove the defensive `max(0, _)` clamp in `LiquidityPoolAggregator.ts:444–457` once the root cause is fixed; or, if kept, log a warning so future drifts are detected
- [ ] Verify the 166-pool count drops to 0 on a full reindex

## Out of scope

- The negative *unstaked* `reserve0`/`reserve1` on Superseed pools — separate root cause, separate issue.
- The `priceFlag` / denylist work for warped USD totals — separate issue.

### PR DESCRIPTION
**Pull request by FonsSicca**
## Summary

- Replaces the post-crossing proportional split of swap reserve deltas (the source of the negative `stakedReserve0/1` drift on 166 CL pools) with **exact per-segment Uniswap v3 math** walking the same staked-tick edge sweep already used to keep `stakedLiquidityInRange` in sync.
- `processTickCrossingsForStaked` now returns `{ stakedLiquidityInRange, stakedDelta0, stakedDelta1 }` and accepts `oldSqrtPriceX96 / newSqrtPriceX96` so attribution and the sweep share one pass. `computeStakedSwapReserveDelta` is removed.
- `L_total` cancels in the staked-share formula (`Δ = L_staked * (...)`), so this needs **no new edge state** beyond the existing `stakedTickEdges` / `stakedTickEdgeNets` parallel arrays — purely a math-correctness fix on top of #649.

Per segment between two crossed staked-tick edges, with constant `L_staked`:

```
Δ0 = L_staked * (S_start - S_end) * Q96 / (S_start * S_end)
Δ1 = L_staked * (S_end - S_start) / Q96
```

Both are signed and direction-agnostic (UP swap → Δ0 negative, Δ1 positive; DOWN flips signs), matching the `event.params.amount0/amount1` sign convention of `CLPool.Swap`.

Closes the AFK Slice 1 of #666.

## Test plan

- [x] `tsc --noEmit` clean
- [x] `pnpm qa --write` clean
- [x] `pnpm test` — 97/97 test files pass
- [x] New unit coverage in `test/Aggregators/CLStakedLiquidity.test.ts`:
  - zero-staked short-circuit returns `{ 0n, 0n, 0n }`
  - in-range single-segment swap matches `calculatePositionAmountsFromLiquidity` diff (±1 wei)
  - position exits range mid-swap — only the in-range segment is attributed
  - position enters range mid-swap
  - position never enters range — zero deltas
  - telescoping invariant: deposit + per-segment swap + withdraw ≈ 0n (±2 wei tolerance for bigint truncation)
  - pre-Initialize guard: returns zeros when `oldSqrtPriceX96 === 0n`
- [x] Updated edge-sanity suite (`CLStakedLiquidityEdgeSanity.test.ts`) for the new return shape — including UP/DOWN round-trip parity over the same window.
- [x] **Slice 2 (next PR / follow-up)**: local end-to-end correctness test against the historical repro pool `10-0x1ea36eb...` (the ~80% TVL outlier flagged in #666) using the existing `repro/666-narrow-config` worktree. Two invariants at the three known checkpoint blocks `124360969`, `124362052`, `124362406`:
  1. **No-negative**: `stakedReserve0/1 >= 0` at every checkpoint.
  2. **On-chain truth**: `stakedReserve0/1` agree with the on-chain `Pool.staked0()/staked1()` reads at each block (within rounding tolerance) — i.e., we're not just preventing the negative, we're producing the *correct* value.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed reserve calculation accuracy for staked positions during swap operations, ensuring proper tracking of staked liquidity amounts across price movements and tick crossings.

* **Tests**
  * Enhanced test coverage for staked liquidity calculations, including edge cases, boundary conditions, and multi-step deposit-swap-withdraw scenarios.

### CODE REVIEW

**Review by WooSungD — APPROVED** [review]

File: src/Aggregators/CLStakedLiquidity.ts [hunk 1]
```diff
@@ -246,29 +316,43 @@ export function applyStakedPositionToEdges(
...
+  stakedDelta0: bigint;
+  stakedDelta1: bigint;
+} {
+  if (tickSpacing === 0n || oldSqrtPriceX96 === 0n || newSqrtPriceX96 === 0n) {
+    return {
+      stakedLiquidityInRange: currentStakedLiqInRange,
+      stakedDelta0: 0n,
+      stakedDelta1: 0n,
+    };
   }
```
[the remark this answers is not in the document]
Comment (FonsSicca) [reply src/Aggregators/CLStakedLiquidity.ts]: Addressed in d271222 — reordered the guards so the tick-range check (and STAKED_TICK_DRIFT log) runs before the zero-sqrt/tickSpacing bailout. Updated the comment to reflect the new ordering rationale.

File: test/Aggregators/CLStakedLiquidity.test.ts [hunk 2]
```diff
@@ -1,14 +1,27 @@
...
 } from "../../src/Aggregators/CLStakedLiquidity";
+import { calculatePositionAmountsFromLiquidity } from "../../src/Helpers";
+
+/**
+ * Returns sqrt(1.0001^tick) in Q64.96 fixed-point as bigint — the same
+ * conversion the source-of-truth helper uses internally. Centralizes the
+ * JSBI→bigint plumbing so test math reads cleanly.
+ */
+const sqrtAt = (tick: bigint): bigint =>
+  BigInt(TickMath.getSqrtRatioAtTick(Number(tick)).toString());
```
[the remark this answers is not in the document]
Comment (FonsSicca) [reply test/Aggregators/CLStakedLiquidity.test.ts]: Addressed in d271222 — extracted sqrtAt() into test/Aggregators/common.ts (with JSDoc) and imported it from both CLStakedLiquidity.test.ts and CLStakedLiquidityEdgeSanity.test.ts.

**Comment by FonsSicca** [comment]
Addressed all three CodeRabbit findings in d271222:

1. **Guard reorder** (`src/Aggregators/CLStakedLiquidity.ts`) — tick-range check now runs before the zero-sqrt/tickSpacing bailout so STAKED_TICK_DRIFT logging fires on uninitialized pools too.
2. **Short-circuit guard tests** (`test/Aggregators/CLStakedLiquidity.test.ts:505-545`, the outside-diff finding) — seeded `currentStakedLiqInRange=0n` to match the precondition that `hasStakes=false` / empty-edges imply no staked liquidity, and added explicit `stakedDelta0/1 === 0n` assertions so the final-segment fallback can't silently emit attribution.
3. **Shared `sqrtAt` helper** — extracted to `test/Aggregators/common.ts` and imported in both `CLStakedLiquidity.test.ts` and `CLStakedLiquidityEdgeSanity.test.ts`.

Also rebased onto current `main` (cleanly absorbs `4d253e1` cross-chain prefetch and `3ff360f` ION/Lisk blacklist — no conflicts with the CL files in this PR).

`pnpm qa --write` clean, `tsc --noEmit` clean, all 98 test files pass locally.
